concurrency_test.go

```diff
@@ -0,0 +1,205 @@
+				Events          []Event         `json:"events"`
+				StatsigMetadata statsigMetadata `json:"statsigMetadata"`
+			}
+			input := &requestInput{}
+			defer req.Body.Close()
+			buf := new(bytes.Buffer)
+			buf.ReadFrom(req.Body)
+
+			json.Unmarshal(buf.Bytes(), &input)
+			flushedEvents = input.Events
+		} else if strings.Contains(req.URL.Path, "get_id_lists") {
+			baseURL := "http://" + req.Host
+			r := map[string]idList{
+				"list_1": {Name: "list_1", Size: 3, URL: baseURL + "/list_1", CreationTime: 1, FileID: "file_id_1"},
+				"list_2": {Name: "list_2", Size: 3, URL: baseURL + "/list_2", CreationTime: 1, FileID: "file_id_2"},
+			}
+			v, _ := json.Marshal(r)
+			res.Write(v)
+		} else if strings.Contains(req.URL.Path, "list_1") {
+			res.Write([]byte("+7/rrkvF6\n"))
+		}
+	}))
+
+	defer testServer.Close()
+	options := &Options{
+		API: testServer.URL,
+		Environment: Environment{
+			Params: map[string]string{
+				"foo": "bar",
+			},
+			Tier: "awesome_land",
+		},
+	}
+
+	InitializeWithOptions("secret-key", options)
+
+	const (
+		goroutines = 10
+		loops      = 10
+	)
+
+	var wg sync.WaitGroup
+	wg.Add(goroutines)
+	for g := 0; g < goroutines; g++ {
+		go func() {
+			defer wg.Done()
+			statsigUser := User{
+				UserID:            fmt.Sprintf("statsig_u_%d", g),
+				Email:             "u@statsig.com",
+				PrivateAttributes: map[string]interface{}{"private": "shh"},
+				Custom:            map[string]interface{}{"key": "value"},
+				CustomIDs:         map[string]string{"randomID": "123456"},
+			}
+			user := User{
+				UserID: "regular_user_id",
+				Email:  "u@gmail.com",
+			}
+			for i := 0; i < loops; i++ {
+				LogEvent(Event{EventName: "test_event", User: user})
+				if CheckGate(statsigUser, "on_for_id_list") || !CheckGate(user, "on_for_id_list") {
+					t.Error("statsigUser should fail and regular user should pass for on_for_id_list gate")
+				}
+
+				if !CheckGate(statsigUser, "on_for_statsig_email") || CheckGate(user, "on_for_statsig_email") {
+					t.Error("statsig user should pass statsig email gate and regular user should fail")
+				}
+
+				exp := GetExperiment(statsigUser, "sample_experiment")
+				if !exp.GetBool("layer_param", false) {
+					t.Error("sample_experiment layer_param not correct")
+				}
+				config := GetConfig(statsigUser, "test_config")
+				if config.GetNumber("number", 420) != 7 {
+					t.Error("test_config number not correct")
+				}
+				layer := GetLayer(statsigUser, "a_layer")
+				if !layer.GetBool("layer_param", false) {
+					t.Error("sample_experiment layer_param not correct")
+				}
+				LogEvent(Event{EventName: "test_event_2", User: statsigUser})
+			}
+		}()
+	}
+	wg.Wait()
+
+	// 10 go routines x 10 loops each x 9 events (2 log event + 7 exposure events) = 900
+	if len(instance.logger.events) != 900 {
+		t.Error("Some events were dropped when calling APIs concurrently")
+	}
+
+	Shutdown()
+	if len(flushedEvents) != 900 {
+		t.Error("Not all events were flushed at shutdown")
```

Comment:
This is a great test - can we also expand it to go past the flush size/interval since the logger's flush() concurrency is not tested here

Reply:
Good point, added more logs to make sure logs are cut and flushed correctly without data loss